File: ros2_fundamentals/create_library_with_header/publisher_library/src/timed_publisher.cpp
#include "publisher_library/timed_publisher.hpp"

namespace publisher_library
{

TimedPublisher::TimedPublisher(
    rclcpp::Node::SharedPtr node,
    const std::string & topic,
    double period_seconds)
: node_(node), message_("Default message"), count_(0)
{
    publisher_ = node_->create_publisher<std_msgs::msg::String>(topic, 10);

    timer_ = node_->create_wall_timer(
        std::chrono::duration<double>(period_seconds),
        std::bind(&TimedPublisher::timerCallback, this)
    );

    RCLCPP_INFO(node_->get_logger(),
        "TimedPublisher created on topic '%s' with period %.2fs",
        topic.c_str(), period_seconds);
}

void TimedPublisher::start()
{
    if (timer_->is_canceled()) {
        timer_->reset();
        RCLCPP_INFO(node_->get_logger(), "TimedPublisher started !");
    }
}

void TimedPublisher::stop()
{
    timer_->cancel();
    RCLCPP_INFO(node_->get_logger(), "TimedPublisher stopped !");
}

void TimedPublisher::setMessage(const std::string & message)
{
    message_ = message;
    RCLCPP_INFO(node_->get_logger(), "Message changed to: '%s'", message_.c_str());
}

void TimedPublisher::timerCallback()
{
    auto msg = std_msgs::msg::String();
    msg.data = message_ + " [" + std::to_string(count_++) + "]";
    publisher_->publish(msg);
    RCLCPP_INFO(node_->get_logger(), "Publishing: '%s'", msg.data.c_str());
}

}  // namespace publisher_library

File: ros2_fundamentals/create_library_with_header_py/publisher_library_py/publisher_library_py/timed_publisher.py
from rclpy.node import Node
from std_msgs.msg import String

# ============================================================
# CLASSE DE LA LIBRAIRIE Python
# ============================================================

class TimedPublisher:
    def __init__(self, node: Node, topic: str, period_seconds: float):
        """
        node           : nœud ROS2 existant
        topic          : nom du topic
        period_seconds : période du timer en secondes
        """
        self.node_    = node
        self.message_ = "Default message"
        self.count_   = 0

        # Créer le publisher sur le nœud fourni
        self.publisher_ = node.create_publisher(String, topic, 10)

        # Créer le timer
        self.timer_ = node.create_timer(period_seconds, self.timer_callback)

        node.get_logger().info(
            f"TimedPublisher created on topic '{topic}' "
            f"with period {period_seconds}s"
        )

    def start(self):
        """Démarrer la publication"""
        if self.timer_.is_canceled():
            self.timer_.reset()
            self.node_.get_logger().info("TimedPublisher started !")

    def stop(self):
        """Arrêter la publication"""
        self.timer_.cancel()
        self.node_.get_logger().info("TimedPublisher stopped !")

    def set_message(self, message: str):
        """Changer le message publié"""
        self.message_ = message
        self.node_.get_logger().info(f"Message changed to: '{message}'")

    def timer_callback(self):
        msg = String()
        msg.data = f"{self.message_} [{self.count_}]"
        self.count_ += 1
        self.publisher_.publish(msg)
        self.node_.get_logger().info(f"Publishing: '{msg.data}'")